File: mobile_sejourfr/lib/core/widgets/paywall_context.dart
/// Ce que l'offre et l'onboarding savent de la date d'examen — règles
/// **pures**, déclarées une fois pour tout le mobile.
///
/// 🛑 **Le paywall ne dit PLUS l'échéance** (demande du propriétaire,
/// 2026-09-20). Après l'en-tête personnalisé, les deux derniers bandeaux sont
/// partis — « Objectif B2 avant le … — il vous reste N jours. » et « Votre
/// examen est le … Le pass N couvre toute votre préparation. » — et avec eux
/// `PaywallContext`, `paywallContext`, `echeanceLine` et `passRecommande`, qui
/// ne servaient plus qu'à les composer. Ne pas les réintroduire : l'offre dit
/// le prix, pas l'échéance.
///
/// ⚠️ **Pas de miroir web pour tout ce fichier.** `web_sejoufr/lib/paywall-context.ts`
/// est **supprimé** — côté web, plus rien ne lisait ces règles. Ce qui reste
/// ici a ses propres lecteurs : [joursAvantExamen] et [formatJour] servent la
/// carte de date d'examen de l'onboarding (`target_path_screen.dart`), et
/// [passFromPrice] est le miroir de `passFromPrice` (`lib/passes.ts`).
library;

import '../models/billing_models.dart';

/// Jours restants avant l'examen. Date passée ⇒ `null`.
int? joursAvantExamen(DateTime? examDate, {DateTime? now}) {
  if (examDate == null) return null;
  final maintenant = now ?? DateTime.now();
  final jours = (examDate.difference(maintenant).inMinutes / (60 * 24)).ceil();
  return jours >= 0 ? jours : null;
}

/// **Le prix d'entrée d'un module** — le plus petit montant réellement vendu.
///
/// 🛑 **Dérivé du catalogue, jamais écrit.** C'est ce que rend « À partir de
/// X € » sur l'écran de déblocage du Plan : un chiffre en dur y vieillirait à
/// la première grille de prix. `null` quand le module ne vend rien (catalogue
/// injoignable, ou aucun pass actif) — l'écran n'affiche alors **aucune** ligne
/// de prix plutôt qu'un montant de repli.
///
/// Miroir de `passFromPrice` (`web_sejoufr/lib/passes.ts`).
double? passFromPrice(
  List<PlanPublicResponse>? plans,
  PlanModuleTarget module,
) =>
    passFromPlan(plans, module)?.price;

/// **Le pass dont [passFromPrice] affiche le prix** — le moins cher du module.
/// Sert à dire, dans la mesure du clic « Débloquer mon plan », quel pass et
/// quel montant le candidat avait sous les yeux. `null` dans les mêmes cas.
PlanPublicResponse? passFromPlan(
  List<PlanPublicResponse>? plans,
  PlanModuleTarget module,
) {
  if (plans == null) return null;
  final cible = module == PlanModuleTarget.civique
      ? ModuleAccess.civique
      : ModuleAccess.integral;
  PlanPublicResponse? min;
  for (final p in plans) {
    if (!p.isOneTime || p.moduleAccess != cible) continue;
    if (min == null || p.price < min.price) min = p;
  }
  return min;
}

/// « 18 octobre ». Le jour tel qu'il a été déclaré, sans fuseau appliqué.
String formatJour(DateTime d) {
  const mois = [
    'janvier', 'février', 'mars', 'avril', 'mai', 'juin',
    'juillet', 'août', 'septembre', 'octobre', 'novembre', 'décembre',
  ];
  return '${d.day} ${mois[d.month - 1]}';
}

File: web_sejoufr/lib/paywall-context.ts
/**
 * Ce que l'offre sait de l'échéance du candidat — règles **pures**, déclarées
 * une fois pour tout le web.
 *
 * 🛑 **L'en-tête personnalisé a été SUPPRIMÉ** (demande du propriétaire,
 * 2026-09-20). « Votre plan B2 est prêt », le pitch, les trois priorités
 * réelles, les bénéfices du plan et `PaywallOrigin` — qui ne servait qu'à
 * décider de leur affichage — sont partis **avec leur seul lecteur** : la
 * promesse est désormais dite **une fois**, sur l'écran de transition
 * `/plan/debloquer`, et la répéter sur la page suivante la disait deux fois de
 * suite. Ne pas les réintroduire ici.
 *
 * Ce qui reste ne parle **pas du plan** : la date d'examen déclarée et le pass
 * le plus court qui la couvre. Ce sont les deux faits qui aident à **choisir
 * une durée**, et c'est bien le rôle d'une page d'offre.
 *
 * 🛑 **Rien n'est dérivé ici.** L'objectif arrive servi ; ce fichier ne fait que
 * le mettre en mots.
 *
 * Miroir de `mobile_sejourfr/lib/core/widgets/paywall_context.dart`.
 */
import type {
    AuthenticatedUser,
    LearningPlanDto,
    PlanPublicResponse,
    TargetLevel,
} from "./types";
import {passDurationLabel} from "./passes";

export interface PaywallContext {
    /** Palier visé. `null` = démarche non déclarée. */
    objectiveLevel: TargetLevel | null;
    /** Jour de l'examen déclaré. `null` = pas de date, réponse pleine. */
    examDate: string | null;
}

/**
 * Assemble le contexte à partir de ce qui est **déjà chargé**.
 *
 * `plan` vient du cache : l'offre ne déclenche aucun appel réseau de plus, et
 * son absence est un cas normal (l'écran qui l'ouvre n'a pas forcément lu le
 * Plan).
 */
export function paywallContext(
    plan: LearningPlanDto | null | undefined,
    user: Pick<AuthenticatedUser, "examDate"> | null | undefined,
): PaywallContext {
    return {
        objectiveLevel: plan?.cycle?.objectiveLevel ?? null,
        examDate: user?.examDate ?? null,
    };
}

/** Jours restants avant l'examen. Négatif ⇒ `null` : la date est passée. */
export function joursAvantExamen(
    examDate: string | null,
    now: Date = new Date(),
): number | null {
    if (!examDate) return null;
    const jour = new Date(`${examDate}T00:00:00`);
    if (Number.isNaN(jour.getTime())) return null;
    const jours = Math.ceil(
        (jour.getTime() - now.getTime()) / (24 * 60 * 60 * 1000),
    );
    return jours >= 0 ? jours : null;
}

/** « Objectif B2 avant le 18 octobre — il vous reste 39 jours. » */
export function echeanceLine(
    ctx: PaywallContext,
    now: Date = new Date(),
): string | null {
    const jours = joursAvantExamen(ctx.examDate, now);
    if (jours === null) return null;
    const date = formatJour(ctx.examDate!);
    const objectif = ctx.objectiveLevel ? `Objectif ${ctx.objectiveLevel} avant` : "Examen";
    const reste =
        jours === 0
            ? "c'est aujourd'hui."
            : `il vous reste ${jours} jour${jours > 1 ? "s" : ""}.`;
    return `${objectif} le ${date} — ${reste}`;
}

/**
 * Le levier propre aux pass (`50_` §3.4) : **aligner la durée sur l'échéance**.
 *
 * « Votre examen est le 18 octobre. Le pass 2 mois couvre toute votre
 * préparation. »
 *
 * 🛑 On ne recommande **que** ce que le catalogue propose réellement, et
 * seulement un pass qui **couvre** l'échéance : promettre une couverture qu'un
 * pass ne tient pas serait pire que se taire. Aucun pass assez long ⇒ `null`.
 */
export function passRecommande(
    ctx: PaywallContext,
    plans: PlanPublicResponse[] | null | undefined,
    now: Date = new Date(),
): {plan: PlanPublicResponse; phrase: string} | null {
    const jours = joursAvantExamen(ctx.examDate, now);
    if (jours === null || !plans?.length) return null;

    // Le plus COURT des pass qui couvrent l'échéance : on ne pousse pas au plus
    // cher, on répond au besoin.
    const couvrants = plans
        .filter((p) => p.purchaseType === "ONE_TIME" && p.durationDays >= jours)
        .sort((a, b) => a.durationDays - b.durationDays);

    const choisi = couvrants[0];
    if (!choisi) return null;

    return {
        plan: choisi,
        phrase: `Votre examen est le ${formatJour(ctx.examDate!)}. Le pass ${passDurationLabel(
            choisi.durationDays,
        )} couvre toute votre préparation.`,
    };
}

/** « 18 octobre ». Le jour tel qu'il a été déclaré, sans fuseau appliqué. */
function formatJour(iso: string): string {
    const [, mois, jour] = iso.split("-");
    const nom = [
        "janvier", "février", "mars", "avril", "mai", "juin",
        "juillet", "août", "septembre", "octobre", "novembre", "décembre",
    ][Number(mois) - 1];
    return nom ? `${Number(jour)} ${nom}` : iso;
}
